File: iOS/MyApplication/Data Sources/CapsulesDataSource.swift
//
//  CapsulesDataSource.swift
//  MyApplication
//
//  Created by ESLealAl on 20/3/23.
//

import Foundation
import FirebaseDatabase

protocol ICapsulesDataSource {
    func get(id: String) -> Capsule?
    func fetch() async throws -> [Capsule]
    func subscribe(callback: @escaping ([Capsule]) -> Void)
}

class CapsulesDataSource: ICapsulesDataSource {
    
    private var capsules: [Capsule] = []

    lazy private var database: DatabaseReference = Database.database().reference().child("capsules")

    func subscribe(callback: @escaping ([Capsule]) -> Void) {
        self.database.observe(.value) { snapshot in
            var fetchedCapsules = [Capsule]()
            for capsuleSnapshot in snapshot.children.allObjects as! [DataSnapshot] {
                fetchedCapsules.append(Capsule(capsuleSnapshot))
            }
            // Updating local copy
            self.capsules = fetchedCapsules
            callback(fetchedCapsules)
        }
        
    }
    
    func fetch() async throws -> [Capsule] {
        return try await withCheckedThrowingContinuation { continuation in
            self.database.observeSingleEvent(of: .value) { snapshot in
                var fetchedCapsules = [Capsule]()
                for capsuleSnapshot in snapshot.children.allObjects as! [DataSnapshot] {
                    fetchedCapsules.append(Capsule(capsuleSnapshot))
                }

                // Updating local copy
                self.capsules = fetchedCapsules
                continuation.resume(returning: fetchedCapsules)
            } withCancel: { error in
                continuation.resume(throwing: error)
            }
        }
    }

    func get(id: String) -> Capsule? {
        return capsules.first(where: { $0.id == id })
    }
}

File: Android/app/src/main/java/com/albertleal/gibernat/myapplication/datasources/CapsulesDataSource.kt
package com.albertleal.gibernat.myapplication.datasources

import com.albertleal.gibernat.myapplication.datasources.interfaces.ICapsulesDataSource
import com.albertleal.gibernat.myapplication.models.Capsule
import com.google.firebase.database.ChildEventListener
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.google.protobuf.Value
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

class CapsulesDataSource(private val database: FirebaseDatabase) : ICapsulesDataSource {

    private var capsules: List<Capsule> = mutableListOf<Capsule>()

    fun subscribe(callback: (List<Capsule>) -> Unit)  {
        val ref = database.getReference("capsules")
        ref.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                val fetchedCapsules = mutableListOf<Capsule>()

                for (capsuleSnapshot in snapshot.children) {
                    val capsule = capsuleSnapshot.getValue(Capsule::class.java)
                    if (capsule != null) {
                        capsule.id = capsuleSnapshot.key
                        fetchedCapsules.add(capsule)
                    }
                }

                //Updating local copy
                capsules = fetchedCapsules
                callback(fetchedCapsules)
            }

            override fun onCancelled(error: DatabaseError) {
                callback(emptyList()) // or callback with some default value
            }
        })
    }

    override suspend fun fetch(): List<Capsule> {
        return suspendCoroutine { continuation ->

            val ref = database.getReference("capsules")
            ref.addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(snapshot: DataSnapshot) {
                    val fetchedCapsules = mutableListOf<Capsule>()

                    for (capsuleSnapshot in snapshot.children) {
                        val capsule = capsuleSnapshot.getValue(Capsule::class.java)
                        if (capsule != null) {
                            capsule.id = capsuleSnapshot.key
                            fetchedCapsules.add(capsule)
                        }
                    }

                    //Updating local copy
                    capsules = fetchedCapsules

                    continuation.resume(fetchedCapsules)
                }

                override fun onCancelled(error: DatabaseError) {
                    continuation.resumeWithException(error.toException())
                }
            })
        }
    }

    override fun get(id: String): Capsule? {
        return capsules.find { it.id == id }
    }

}